Include/Rover/Range.hpp

```diff
@@ -75,11 +76,90 @@ namespace Rover {
         Interval interval = Interval::CLOSED);
 
       Type generate(Evaluator& evaluator);
+
+    private:
+      Begin m_begin;
+      End m_end;
+      std::function<Type(const Type&)> m_round;
+      Interval m_interval;
+      std::mt19937 m_engine;
+      std::uniform_real_distribution<double> m_distribution;
   };
 
   template<typename BeginFwd, typename EndFwd>
   Range(BeginFwd&&, EndFwd&&, Interval = Interval::CLOSED) ->
     Range<std::decay_t<BeginFwd>, std::decay_t<EndFwd>, void>;
+
+  template<typename BeginFwd, typename EndFwd, typename GranularityFwd>
+  Range(BeginFwd&&, EndFwd&&, GranularityFwd&&, Interval = Interval::CLOSED) ->
+    Range<std::decay_t<BeginFwd>, std::decay_t<EndFwd>,
+    std::decay_t<GranularityFwd>>;
+
+  template<typename B, typename E, typename G>
+  template<typename BeginFwd, typename EndFwd>
+  Range<B, E, G>::Range(BeginFwd&& begin, EndFwd&& end,
+      Interval interval)
+      : m_begin(std::forward<BeginFwd>(begin)),
+        m_end(std::forward<EndFwd>(end)),
+        m_round([](const Type& value){
+          return value;
+        }),
+        m_interval(interval),
+        m_engine(std::random_device()()),
+        m_distribution(0., 1.) {}
+
+  template<typename B, typename E, typename G>
+  template<typename BeginFwd, typename EndFwd, typename GranularityFwd>
+  Range<B, E, G>::Range(BeginFwd&& begin, EndFwd&& end,
+      GranularityFwd&& granularity, Interval interval)
+      : m_begin(std::forward<BeginFwd>(begin)),
+        m_end(std::forward<EndFwd>(end)),
+        m_round([granularity = std::forward<GranularityFwd(granularity)](
+            const Type& value){
+          auto flr = granularity * static_cast<int64_t>(value / granularity);
+          auto ceil = granularity * (static_cast<int64_t>(value /
+              granularity) + 1);
+          return std::abs(value - flr) <= std::abs(ceil - value) ?
+            flr : ceil;
+        }),
+        m_interval(interval),
+        m_engine(std::random_device()()),
+        m_distribution(0., 1.) {}
+
+  template<typename B, typename E, typename G>
+  typename Range<B, E, G>::Type Range<B, E, G>::generate(
+      Evaluator& evaluator) {
+    auto begin = evaluator.evaluate(m_begin);
+    auto end = evaluator.evaluate(m_end);
+    while(true) {
+      auto random_fraction = m_distribution(m_engine);
```

Comment:
I did not find a benefit of using uniform_int_distribution from -INF to +INF over the real distribution from 0 to 1, because I would have to calculate a floating point fraction like this anyway, or use even more intrusive arithmetics on generators (i.e. to each generator would have to be either "integer" or "real"). Let me know if I am missing something.